test/box-luatest/gh_3211_per_module_log_level_test.lua

```diff
@@ -59,3 +67,122 @@ g1.test_print_module_name = function(cg)
+
+        t.assert_equals(log.cfg.level_modules, nil)
+        t.assert_equals(box.cfg.log_level_modules, nil)
+        log_cfg{level = 'info',
+                level_modules = {mod1 = 'debug',
+                                 ['mod2'] = 2,
+                                 ['mod3'] = 'error'}}
+        -- Check that log.cfg{} without 'level_modules' options doesn't affect
+        -- per-module levels.
+        log_cfg{level = 'warn'}
+        t.assert_equals(log.cfg.level, 'warn')
+        t.assert_equals(log.cfg.level_modules.mod1, 'debug')
+        t.assert_equals(log.cfg.level_modules.mod2, 2)
+        t.assert_equals(log.cfg.level_modules.mod3, 'error')
+        assert_log_and_box_cfg_equals()
+        --[[ TODO(gh-7962)
+        -- Check that log.cfg{level_modules = {...}} with the new modules
+        -- appends them to the existing ones.
+        log_cfg{level_modules = {mod4 = 4}}
+        t.assert_equals(log.cfg.level_modules.mod1, 'debug')
+        t.assert_equals(log.cfg.level_modules.mod2, 2)
+        t.assert_equals(log.cfg.level_modules.mod3, 'error')
+        t.assert_equals(log.cfg.level_modules.mod4, 4)
+        assert_log_and_box_cfg_equals()
+        -- Check that log.cfg{level_modules = {...}} sets levels for the
+        -- specified modules, and doesn't affect other modules.
+        log_cfg{level_modules = {mod1 = 0, mod3 = 'info'}}
+        t.assert_equals(log.cfg.level_modules.mod1, 0)
+        t.assert_equals(log.cfg.level_modules.mod2, 2)
+        t.assert_equals(log.cfg.level_modules.mod3, 'info')
+        t.assert_equals(log.cfg.level_modules.mod4, 4)
+        assert_log_and_box_cfg_equals()
+        -- Check that box.NULL and the empty string remove the corresponding
+        -- module from the config.
+        log_cfg{level_modules = {mod2 = box.NULL, mod4 = ''}}
+        t.assert_equals(log.cfg.level_modules.mod1, 0)
+        t.assert_equals(log.cfg.level_modules.mod2, nil)
+        t.assert_equals(log.cfg.level_modules.mod3, 'info')
+        t.assert_equals(log.cfg.level_modules.mod4, nil)
+        assert_log_and_box_cfg_equals()
+        -- Check that log.cfg{level_modules = box.NULL} removes all modules.
+        log_cfg{level_modules = box.NULL}
+        t.assert_equals(log.cfg.level_modules, nil)
+        t.assert_equals(box.cfg.log_level_modules, nil)
+        --]]
+        -- Check that log levels actually affect what is printed to the log.
+        local log1 = log.new('mod1')
+        local log2 = log.new('mod2')
+        log_cfg{level = 'info', level_modules = {mod1 = 4, mod2 = 'debug'}}
+        log.warn('warn from the default log')
+        log.info('info from the default log')
+        log.debug('debug from the default log')
+        log1.warn('warn from mod1')
+        log1.info('info from mod1')
+        log1.debug('debug from mod1')
+        log2.warn('warn from mod2')
+        log2.info('info from mod2')
+        log2.debug('debug from mod2')
```

Comment:
Let's also check that if level for logger is not set in `level_modules` then it is global one.

Reply:
Done